lib/kfuzztest/relocations.c

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright 2025 Google LLC */
+#include <linux/kfuzztest.h>
+#include <linux/kasan.h>
+
+/**
+ * Poison the half open interval [start, end], where end should be 8-byte 
+ * aligned if it is not, then we cannot guarantee that the whole range will
+ * be poisoned.
+ *
+ * If start is not 8-byte-aligned, the remaining bytes in its 8-byte granule
+ * can only be poisoned if CONFIG_KASAN_GENERIC is enabled.
+ */
+static void kfuzztest_poison_range(void *start, void *end)
+{
+	uintptr_t end_addr = ALIGN_DOWN((uintptr_t)end, __KASAN_GRANULE_SIZE);
+	uintptr_t start_addr = (uintptr_t)start;
+	uintptr_t head_granule_start;
+	uintptr_t poison_body_start;
+	uintptr_t poison_body_end;
+	size_t head_prefix_size;
+
+	if (!IS_ENABLED(CONFIG_KASAN))
+		return;
+
+	if (start_addr >= end_addr)
+		return;
+
+	head_granule_start = ALIGN_DOWN(start_addr, __KASAN_GRANULE_SIZE);
+	head_prefix_size = start_addr - head_granule_start;
+
+	if (IS_ENABLED(CONFIG_KASAN_GENERIC) && head_prefix_size > 0)
+		kasan_poison_last_granule((void *)head_granule_start,
+					  head_prefix_size);
+
+	poison_body_start = ALIGN(start_addr, __KASAN_GRANULE_SIZE);
+	poison_body_end = ALIGN_DOWN(end_addr, __KASAN_GRANULE_SIZE);
+
+	if (poison_body_start < poison_body_end)
+		kasan_poison((void *)poison_body_start,
+			     poison_body_end - poison_body_start,
+			     __KASAN_SLAB_REDZONE, false);
+}
+
+int __kfuzztest_relocate(struct reloc_region_array *regions,
+			 struct reloc_table *rt, void *payload_start,
+			 void *payload_end)
+{
+	struct reloc_region reg, src, dst;
+	void *poison_start, *poison_end;
+	uintptr_t *ptr_location;
+	struct reloc_entry re;
+	size_t i;
+
+	/* Patch pointers. */
+	for (i = 0; i < rt->num_entries; i++) {
+		re = rt->entries[i];
+		src = regions->regions[re.region_id];
+		ptr_location = (uintptr_t *)((char *)payload_start +
+					     src.offset + re.region_offset);
+		if (re.value == KFUZZTEST_REGIONID_NULL)
+			*ptr_location = (uintptr_t)NULL;
+		else if (re.value < regions->num_regions) {
+			dst = regions->regions[re.value];
+			*ptr_location =
+				(uintptr_t)((char *)payload_start + dst.offset);
+		} else
+			return -EINVAL;
+	}
+
+	/* Poison the padding between regions. */
```

Comment:
Given that we already do not poison anything if CONFIG_KASAN is disabled, can we also not execute this code?
You can move it to a separate function that is empty or bails out under CONFIG_KASAN.